src/csync/csync_reconcile.cpp

```diff
@@ -316,6 +317,33 @@ static int _csync_merge_algorithm_visitor(csync_file_stat_t *cur, CSYNC * ctx) {
                         (ctx->current == REMOTE_REPLICA ? cur->checksumHeader : other->checksumHeader);
                     if (!remoteChecksumHeader.isEmpty()) {
                         is_conflict = true;
+
+                        // Do we have an UploadInfo for this?
+                        // Maybe the Upload was completed, but the connection was broken just before
+                        // we recieved the etag (Issue #5106)
+                        auto up = ctx->statedb->getUploadInfo(cur->path);
+                        if (up._valid && up._contentChecksum == remoteChecksumHeader) {
+                            // Solve the conflict into an upload, or nothing
+                            auto remoteNode = ctx->current == REMOTE_REPLICA ? cur : other;
+                            auto localNode = ctx->current == REMOTE_REPLICA ? other : cur;
+                            remoteNode->instruction = CSYNC_INSTRUCTION_NONE;
+                            localNode->instruction = up._modtime == localNode->modtime ? CSYNC_INSTRUCTION_UPDATE_METADATA : CSYNC_INSTRUCTION_SYNC;
+
+                            // Update the etag and other server metadata in the journal already
```

Comment:
That's not safe to do in the SYNC case, is it?

Reply:
I think it is.  Why wouldn't it?

Reply:
So, assume we end up in the sync case. That means we have an upload info with the same checksum as the remote, but the mtime differs. We set a SYNC instruction to trigger the download but also immediately adjust the database entry to have the new remote metadata.

Say the sync is aborted before the propagation of that SYNC actually happens. Would the following sync run still do the right thing given that the metadata in the db has already been updated? I expect it would not trigger the SYNC again?

Reply:
Thanks for the explanation!